hibernate-core/src/main/java/org/hibernate/query/sqm/internal/AbstractSqmSelectionQuery.java

```diff
@@ -73,8 +99,164 @@ protected void errorOrLogForPaginationWithCollectionFetch() {
+		return this;
+	}
+
+	@Override
+	public SelectionQuery<R> setPage(KeyedPage<R> page) {
+		keyedPage = page;
+		return this;
+	}
+
+	static class KeyedResult<R> {
+		final R result;
+		final List<Comparable<?>> key;
+		public KeyedResult(R result, List<Comparable<?>> key) {
+			this.result = result;
+			this.key = key;
+		}
+
+		public R getResult() {
+			return result;
+		}
+
+		public List<Comparable<?>> getKey() {
+			return key;
+		}
+	}
+
+	private SqmSelectStatement<KeyedResult<R>> keyed(List<Order<? super R>> keyDefinition, List<Comparable<?>> keyValues) {
+		@SuppressWarnings("unchecked")
+		final SqmSelectStatement<KeyedResult<R>> sqm =
+				(SqmSelectStatement<KeyedResult<R>>)
+						getSqmSelectStatement().copy( noParamCopyContext() );
+		final NodeBuilder builder = sqm.nodeBuilder();
+		final SqmQuerySpec<?> querySpec = sqm.getQuerySpec();
+		final SqmWhereClause whereClause = querySpec.getWhereClause();
+		final List<? extends JpaSelection<?>> items = querySpec.getSelectClause().getSelectionItems();
+		final List<SqmPath<?>> newItems = new ArrayList<>();
+		if ( items.size() == 1 ) {
+			final JpaSelection<?> selected = items.get(0);
+			for ( int i = 0; i < keyDefinition.size(); i++ ) {
+				// ordering by an attribute of the returned entity
+				if ( selected instanceof SqmRoot ) {
+					final Order<? super R> key = keyDefinition.get(i);
+					final SqmFrom<?,?> root = (SqmFrom<?,?>) selected;
+					if ( keyValues != null ) {
+						@SuppressWarnings("rawtypes")
+						final Comparable keyValue = keyValues.get(i);
+						whereClause.applyPredicate( keyPredicate( root, key, keyValue, builder ) );
+					}
+					newItems.add( root.get( key.getAttributeName() ) );
+				}
+				else {
+					throw new IllegalQueryOperationException("Select item was not an entity type");
+				}
+			}
+			sqm.select( builder.construct((Class) KeyedResult.class,
+					asList(selected, builder.construct(List.class, newItems)) ) );
+			return sqm;
+		}
+		else {
+			throw new IllegalQueryOperationException("Query has multiple items in the select list");
+		}
+	}
+
+	private <C extends Comparable<? super C>> SqmPredicate keyPredicate(
+			SqmFrom<?, ?> selected, Order<? super R> key, C keyValue, NodeBuilder builder) {
+		if ( !key.getEntityClass().isAssignableFrom( selected.getJavaType() ) ) {
+			throw new IllegalQueryOperationException("Select item was of wrong entity type");
+		}
+		final Expression<? extends C> path = selected.get( key.getAttributeName() );
+		// TODO: use a parameter here and create a binding for it
+//		@SuppressWarnings("unchecked")
+//		final Class<C> valueClass = (Class<C>) keyValue.getClass();
+//		final JpaParameterExpression<C> parameter = builder.parameter(valueClass);
+//		setParameter( parameter, keyValue );
+		switch ( key.getDirection() ) {
+			case ASCENDING:
+				return builder.greaterThan( path, builder.literal(keyValue) );
+			case DESCENDING:
+				return builder.lessThan( path, builder.literal(keyValue) );
```

Comment:
The following will create a parameter including a binding automatically.

```suggestion
			case ASCENDING:
				return builder.greaterThan( path, keyValue );
			case DESCENDING:
				return builder.lessThan( path, keyValue );
```

Reply:
That's actually the first thing I tried, but it doesn't actually seem to give me a query parameter. Not sure if there's something I need to set.